### ISSUE TITLE
panic: semaphore: bad release

### ISSUE THREAD

**Issue by mritd** [containers/image#901]
I created a tool to synchronize the gcr.io image to the docker hub, but after running for a period of time, I get a `panic: semaphore: bad release`

```sh
INFO[2020-04-21 17:31:14] sync k8s.gcr.io/k8s-tpu-operator:0.1 => docker.io/gcrxio/k8s.gcr.io_k8s-tpu-operator:0.1
INFO[2020-04-21 17:31:14] sync k8s.gcr.io/k8s-tpu-operator:0585ea2d8131291cc041329e32bb7c81263cf7c5 => docker.io/gcrxio/k8s.gcr.io_k8s-tpu-operator:0585ea2d8131291cc041329e32bb7c81263cf7c5
INFO[2020-04-21 17:31:14] sync k8s.gcr.io/kube-addon-manager-ppc64le:v6.4-alpha.3 => docker.io/gcrxio/k8s.gcr.io_kube-addon-manager-ppc64le:v6.4-alpha.3
INFO[2020-04-21 17:31:14] sync k8s.gcr.io/k8s-jupyterhub:0.2 => docker.io/gcrxio/k8s.gcr.io_k8s-jupyterhub:0.2
INFO[2020-04-21 17:31:18] sync k8s.gcr.io/k8s-tpu-operator:1.0 => docker.io/gcrxio/k8s.gcr.io_k8s-tpu-operator:1.0
panic: semaphore: bad release

goroutine 81928 [running]:
golang.org/x/sync/semaphore.(*Weighted).Release(0xc007f61ae0, 0x1)
        /Users/natural/gopath/pkg/mod/golang.org/x/sync@v0.0.0-20190227155943-e225da77a7e6/semaphore/semaphore.go:98 +0x1f5
github.com/containers/image/copy.(*imageCopier).copyLayers.func1(0xb, 0xc00066b180, 0x47, 0x15d3896, 0x0, 0x0, 0x0, 0x0, 0xc00028b900, 0x31, ...)
        /Users/natural/gopath/pkg/mod/github.com/containers/image@v3.0.2+incompatible/copy/copy.go:489 +0x241
created by github.com/containers/image/copy.(*imageCopier).copyLayers.func2
        /Users/natural/gopath/pkg/mod/github.com/containers/image@v3.0.2+incompatible/copy/copy.go:497 +0x1f6
```

For speed, I am performing image synchronization in 20 concurrency, this is my code

```golang
type Image struct {
	Repo string
	User string
	Name string
	Tag  string
}

type SyncOption struct {
	Timeout    time.Duration
	Limit      int
	User       string
	Password   string
	NameSpace  string
	Proxy      string
	QueryLimit int
	Kubeadm    bool
}

func sync2DockerHub(image *Image, opt SyncOption) error {
	destImage := Image{
		Repo: DefaultDockerRepo,
		User: opt.User,
		Name: image.MergeName(),
		Tag:  image.Tag,
	}

	logrus.Infof("sync %s => %s", image, destImage.String())

	ctx, cancel := context.WithTimeout(context.Background(), opt.Timeout)
	defer cancel()

	policyContext, err := signature.NewPolicyContext(
		&signature.Policy{
			Default: []signature.PolicyRequirement{signature.NewPRInsecureAcceptAnything()},
		},
	)
	if err != nil {
		return err
	}
	defer func() { _ = policyContext.Destroy() }()

	srcRef, err := docker.Transport.ParseReference("//" + image.String())
	if err != nil {
		return err
	}
	destRef, err := docker.Transport.ParseReference("//" + destImage.String())
	if err != nil {
		return err
	}

	sourceCtx := &types.SystemContext{DockerAuthConfig: &types.DockerAuthConfig{}}
	destinationCtx := &types.SystemContext{DockerAuthConfig: &types.DockerAuthConfig{
		Username: opt.User,
		Password: opt.Password,
	}}

	m, err := copy.Image(ctx, policyContext, destRef, srcRef, &copy.Options{
		ReportWriter:          ioutil.Discard,
		SourceCtx:             sourceCtx,
		DestinationCtx:        destinationCtx,
		ProgressInterval:      1 * time.Second,
		ForceManifestMIMEType: manifest.DockerV2Schema2MediaType,
	})
	if err != nil {
		return err
	}

	storageDir := filepath.Join(ManifestDir, image.Repo, image.User, image.Name)
	// ignore other error
	if _, err := os.Stat(storageDir); err != nil {
		if err := os.MkdirAll(storageDir, 0755); err != nil {
			return err
		}
	}

	return ioutil.WriteFile(filepath.Join(storageDir, image.Tag+".json"), m, 0644)
}
```

**Comment by mtrmac** [comment]
Thanks for your report.

Can you upgrade to the latest version, please? This looks like #831 .

**Comment by mritd** [comment]
Yes, I have upgraded to the master version, but the problem still exists.
Interestingly, I replaced `ReportWriter: ioutil.Discard` with `ReportWriter: & bytes.Buff` and I get a `panic: * mpb.progress instance can't be reused after it's done!` Error; 
I realized it might be caused by mpb Problem, in the end I changed it to `ReportWriter: os.Stdout`, panic has never appeared...

**Comment by mritd** [comment]
@markmc Is there a way to use this library without using mbp, as a library dependency, the large amount of logs generated by mbp is useless.

**Comment by mritd** [comment]
The embarrassing thing is that pannic appeared again after running about 20,000 times 😳

```sh
panic: *mpb.Progress instance can't be reused after it's done!

goroutine 3043007 [running]:
github.com/vbauerster/mpb/v5.(*Progress).Add(0xc005d55320, 0xffffffffffffffff, 0xca9ca0, 0xc00f234f30, 0xc02ef02400, 0x3, 0x3, 0x20)
        /Users/natural/gopath/pkg/mod/github.com/vbauerster/mpb/v5@v5.0.3/progress.go:139 +0x35f
github.com/vbauerster/mpb/v5.(*Progress).AddSpinner(0xc005d55320, 0xffffffffffffffff, 0x0, 0xc02ef02400, 0x3, 0x3, 0xc00f234ed0)
        /Users/natural/gopath/pkg/mod/github.com/vbauerster/mpb/v5@v5.0.3/progress.go:108 +0xd6
github.com/containers/image/v5/copy.(*copier).createProgressBar(0xc00898ab40, 0xc005d55320, 0xc008407d10, 0x47, 0xffffffffffffffff, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/natural/gopath/pkg/mod/github.com/containers/image/v5@v5.4.3/copy/copy.go:991 +0xa7f
github.com/containers/image/v5/copy.(*imageCopier).copyLayer(0xc003bb2680, 0xcba140, 0xc005f49f20, 0xc008407d10, 0x47, 0xffffffffffffffff, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/natural/gopath/pkg/mod/github.com/containers/image/v5@v5.4.3/copy/copy.go:1058 +0xb42
github.com/containers/image/v5/copy.(*imageCopier).copyLayers.func1(0xb, 0xc008407d10, 0x47, 0xffffffffffffffff, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/natural/gopath/pkg/mod/github.com/containers/image/v5@v5.4.3/copy/copy.go:828 +0x435
created by github.com/containers/image/v5/copy.(*imageCopier).copyLayers.func2
        /Users/natural/gopath/pkg/mod/github.com/containers/image/v5@v5.4.3/copy/copy.go:860 +0x210
```

**Comment by mtrmac** [comment]
Thanks, that looks like another bug on the timeout handling path (we cancel all `copyLayerHelper`s, but don’t wait for them to finish before tearing down the `progressPool`.

**Comment by mtrmac** [comment]
> @markmc Is there a way to use this library without using mpb, as a library dependency, the large amount of logs generated by mbp is useless.

There isn’t a way to completely remove it, but setting `options.ReportWriter` to a non-TTY (including the default of `nil`) should cause the progress bars to be discarded. OTOH, that doesn’t prevent panics due to calling that code incorrectly.

**Comment by mritd** [comment]
I checked the source code, if the variable is nil, it will be discarded by default (ioutil.Discard); 
but the `panic: * mpb.progress instance can't be reused after it's done!` error will occasionally appear, if there is one to completely close mpb the option called will be better.

**Comment by mritd** [comment]
This library seems very unstable, and now there is a new pannic(#750)...

```
panic: d.nx != 0

goroutine 2805597 [running]:
crypto/sha256.(*digest).checkSum(0xc00f134840, 0x0, 0x0, 0x0, 0x0)
        /Users/natural/devtools/go/src/crypto/sha256/sha256.go:234 +0x1d9
crypto/sha256.(*digest).Sum(0xc011f2d200, 0x0, 0x0, 0x0, 0xc00f134968, 0xaf01a765, 0xfead14f1e5cc7dbf)
        /Users/natural/devtools/go/src/crypto/sha256/sha256.go:210 +0x69
github.com/opencontainers/go-digest.NewDigest(0xba9853, 0x6, 0xcc3f20, 0xc011f2d200, 0x101, 0xc00f134ef0)
        /Users/natural/gopath/pkg/mod/github.com/opencontainers/go-digest@v1.0.0-rc1/digest.go:40 +0x46
github.com/opencontainers/go-digest.(*digester).Digest(0xc011f7dc20, 0xcc2520, 0xc00deec0c0)
        /Users/natural/gopath/pkg/mod/github.com/opencontainers/go-digest@v1.0.0-rc1/digester.go:38 +0x49
github.com/containers/image/v5/docker.(*dockerImageDestination).PutBlob(0xc0123caa50, 0xcc2520, 0xc00deec0c0, 0x7f50a889e7d8, 0xc00f149560, 0xc003aac910, 0x47, 0xa4e8d6, 0x0, 0x0, ...)
        /Users/natural/gopath/pkg/mod/github.com/containers/image/v5@v5.4.3/docker/docker_image_dest.go:172 +0x69f
github.com/containers/image/v5/copy.(*copier).copyBlobFromStream(0xc012b64b40, 0xcc2520, 0xc00deec0c0, 0xcb2740, 0xc01055e900, 0xc003aac910, 0x47, 0xa4e8d6, 0x0, 0x0, ...)
        /Users/natural/gopath/pkg/mod/github.com/containers/image/v5@v5.4.3/copy/copy.go:1331 +0x56b
github.com/containers/image/v5/copy.(*imageCopier).copyLayerFromStream(0xc00cc8f800, 0xcc2520, 0xc00deec0c0, 0xcb2740, 0xc01055e900, 0xc003aac910, 0x47, 0xa4e8d6, 0x0, 0x0, ...)
        /Users/natural/gopath/pkg/mod/github.com/containers/image/v5@v5.4.3/copy/copy.go:1131 +0x1e5
github.com/containers/image/v5/copy.(*imageCopier).copyLayer(0xc00cc8f800, 0xcc2520, 0xc00deec0c0, 0xc003aac910, 0x47, 0xa4e8d6, 0x0, 0x0, 0x0, 0x0, ...)
```

**Comment by mtrmac** [comment]
If you can reproduce the panics at will, can you test the combination of #907 and #908 , please?

**Comment by mritd** [comment]
After I added the caching mechanism, panic has rarely appeared; I try to merge this PR for a more detailed test later.

**Comment by CooperLi** [comment]
> After I added the caching mechanism, panic has rarely appeared; I try to merge this PR for a more detailed test later.

@mritd
Is there any progress about **this PR**?
BTW, can't find the PR's link..

**Comment by mritd** [comment]
@CooperLi PR: #907 and #908

**Comment by CooperLi** [comment]
> @CooperLi PR: #907 and #908

@mritd 
appriciate it :-)

**Comment by rhatdan** [comment]
If those are merged, shoudl we close this issue?

**Comment by mtrmac** [comment]
Yes; both PRs have been merged, and c/image 5.4.4, which contains both, has been released.

That should fix the two reported bugs; there may well be more bugs with timeouts/cancellation, please open separate issues for each unique crash.
